ZiLing/Midterm-Review.py

```diff
@@ -0,0 +1,481 @@
+"""
+Midterm review for cs 114
+
+Instructor: Cameron Morland - cjmorland@uwaterloo.ca
+
+This review will cover materials from module 1 to 6, from all the slides 
+and the textbook materials
+
+"""
+
+"""
+Let's start with the very basic
+
+Types         name                        examples 
+
+int          Interger                  3,4,-1,2325289524294
+float   Float-Point Numbers            4.5,0.0,-4305.2458532
+str          String                   'hello', "world", " ",''
+bool         Boolean                   True,False, 1 == 1, 1 != 1
+list          List                     [], [12,3.0432,'hello'],['a','b','v']
+tuple         tuple                    (True,True,False),(),(23,43.2,"Tuple")
+dict        Dictionary                 {2:4},{454:'ok','ok':'No'}
+range         Range                    range(4),range(1,4302,45)
+callable     Callable                  print(),count_To_Four(), f()
+None         NoneType                          None
+Further explination for each will be explained further later, now just note that 
+these are all the types we are working with  
+"""
+
+# You can always use type() to check of the type of certain variable 
+x = 3.23546548432741
+print(type(x))
+
+"""
+Variables
+
+Variable is "something" that has a value inside 
+for example: 
+
+x = 4 
+regdhtweawf__ewgf43 = True 
+name = "value"
+are all variables 
+
+be aware that the variable name cannot contain any special character 
+other than '_' and it cannot start with a number
+"""
+n = 234
+print(n)
+"""
+Function 
+
+a sample function would be print()
+
+All the def balabala(this:that) -> whocares: things you wrote are functions
+function are composed in a few parts 
+
+There are two types of function
+
+Fruitful functions: returns something 
+Void function: returns nothing 
+
+
+here is some function for converting types 
+
+int()     34.2 -> 34
+float()    0 -> 0.0
+str()      23.5 -> "23.5"
+"""
+# Here is a sample function with the name of their different parts 
+# This function WILL NOT ACTUALLY RUN make sure to turn this off when you try to 
+# Run other parts 
+x = 0
+def FunctionName (parameter:type) -> returnType:
+    """
+    Docstring that describes what is the return of this type 
+    
+    Restritoin: restrictions for this function
+    """
+    # Note, innerVariable only exists inside this function
+    # if you call innerVariable outside of the function, error will show 
+    # That this variable does not exist
+    innerVariable = 'something'
+    # But the function can access variables outside of itself
+    x = 1 
+    return 'this is the output of the function'
+
+#when you call a function, you are typing its name and input the parameter
+FunctionName("input")
+
+# You can do math or print multiple things in a single line of print()
+x = 9.81
+print(3.53+5, "hello"+" world",x," m/s^2")
+
+"""
+While we are on the topic of function, lets talk about check 
+
+import check 
```

Comment:
Check has compatibility issues with Python 3.11. If you want to update to the latest version of python, keep your older version of python on your computer for doing assignments that require check. Honestly, you should get in the habit of writing your own tests though since Check is kind of limited.